torch/nn/parallel/distributed.py

```diff
@@ -840,37 +845,30 @@ def forward(self, *inputs, **kwargs):
             else:
                 output = self.module(*inputs, **kwargs)
 
-            if grad_enabled and self.require_backward_grad_sync:
-                self.require_forward_param_sync = True
-                if self.static_graph or not self.find_unused_parameters:
-                    self.reducer.prepare_for_backward([])
-            else:
-                self.require_forward_param_sync = False
-
-        # TODO: DDPSink is currently enabled for unused parameter detection and
-        # static graph training for first iteration, in the future we plan to
-        # enable this passthrough for all training use cases.
-        if (self.find_unused_parameters and not self.static_graph) or (
-            self.static_graph and self.num_iterations == 1
-        ):
-            find_unused = all([
-                grad_enabled,
-                self.require_backward_grad_sync,
-                self.find_unused_parameters,
-            ])
-            state_dict = {
-                'static_graph': self.static_graph,
-                'find_unused': find_unused,
-                'num_iterations': self.num_iterations,
-            }
-            output_tensor_list, treespec = tree_flatten(output)
-            passthrough_tensor_list = _DDPSink.apply(
-                self.reducer,
-                state_dict,
-                *output_tensor_list,
+            self.require_forward_param_sync = (
+                grad_enabled and self.require_backward_grad_sync
             )
-            # Reconstruct output data structure.
-            output = tree_unflatten(passthrough_tensor_list, treespec)
+
+        state_dict = {
+            'static_graph': self.static_graph,
+            'find_unused': self.find_unused_parameters,
+            'num_iterations': self.num_iterations,
```

Comment:
self.num_iterations is calculated for forward pass in python side, may not be synced with num_bwd_calls_ in C++

Reply:
Agreed it is not synced, but we really only use this to decide whether we need to enqueue delay allreduce right? And we only enqueue delay allreduce exactly once (basically only the first call to backward), so just checking num_iterations == 1 should be sufficient as delay allreduce will run in the next bwd.

Although agreed it might be a bit cleaner to directly check the num_backward_calls, perhaps

Reply:
discussed offline, we are using  ctx.reduer.get_num_backward_calls_() as suggested